crossword-xml-uploader/src/main/scala/com/gu/crossword/Lambda.scala

```diff
@@ -1,27 +1,74 @@
 package com.gu.crossword
 
-import java.util.{ Map => JMap }
-import com.amazonaws.services.lambda.runtime.{ RequestHandler, Context }
-import com.gu.crossword.crosswords.{ CrosswordStore, CrosswordUploader }
+import java.util.{Map => JMap}
+import com.amazonaws.services.lambda.runtime.{Context, RequestHandler}
+import com.gu.crossword.crosswords._
+import com.gu.crossword.crosswords.models.CrosswordXmlFile
 
-class Lambda
-    extends RequestHandler[JMap[String, Object], Unit]
-    with CrosswordUploader
-    with CrosswordStore {
+import scala.util.{Failure, Success}
+
+trait CrosswordUploaderLambda
+  extends RequestHandler[JMap[String, Object], Unit]
+    with ComposerOps
+    with CrosswordStore
+    with CrosswordConfigRetriever
+    with CrosswordUploader {
+
+  private def doUpload(url: String, streamName: String, crosswordXmlFile: CrosswordXmlFile): Either[(String, Throwable), String] = {
+    val uploadResult = for {
+      rawXml <- uploadCrossword(url)(crosswordXmlFile)
+      crosswordXml <- XmlProcessor.process(rawXml)
+      _ <- createPage(streamName)(crosswordXmlFile.key, crosswordXml)
+    } yield ()
 
-  override def handleRequest(event: JMap[String, Object], context: Context): Unit = {
+    uploadResult match {
+      case Success(_) => Right(crosswordXmlFile.key)
+      case Failure(error) => Left((crosswordXmlFile.key, error))
+    }
+  }
 
-    implicit val config = new Config(context)
+  def handleRequest(event: JMap[String, Object], context: Context): Unit = {
+    val config = getConfig(context)
 
     println("The uploading of crossword xml files has started.")
 
-    for {
-      crosswordXmlFile <- getCrosswordXmlFiles(config)
-    } yield {
-      uploadCrossword(crosswordXmlFile)
+    val crosswordXmlFiles = getCrosswordXmlFiles(config.crosswordsBucketName)
+    println(s"Found ${crosswordXmlFiles.size} crossword file(s) to process")
+
+    val (failures, successes) = crosswordXmlFiles.map { crosswordXmlFile =>
+      doUpload(
+        url = config.crosswordMicroAppUrl,
+        streamName = config.composerCrosswordIntegrationStreamName,
+        crosswordXmlFile = crosswordXmlFile
+      )
+    } partitionMap(identity)
+
+    failures.foreach {
+      case (key, error) =>
+        println(s"Failed to upload crossword ${key} with error: ${error.getMessage}")
+        error.getStackTrace.foreach(println)
+        archiveFailedCrosswordXMLFile(config.crosswordsBucketName, key)
+    }
+
+    successes.foreach { key =>
+      println(s"Successfully uploaded crossword ${key}")
+      archiveCrosswordXMLFile(config.crosswordsBucketName, key)
     }
 
-    println("The uploading of crossword xml files has finished.")
+    println(s"The uploading of crossword xml files has finished, ${successes.size} succeeded, ${failures.size} failed.}")
+
+    // We want to fail the lambda if any of the uploads failed
```

Comment:
Does this add new behaviour, or does it just make some existing behaviour explicit?

iiuc lambda failure/success can have effects on other AWS services that are connected to them (e.g. whether the condition condition which triggered the lambda is treated as having been resolved or not). So if this is new behaviour, should we document the expectation that this behaviour won't have any unintended side effects like this?

Reply:
My reading is that previously unhandled exceptions would bubble up to the top and cause a lambda failure when errors occurred so in that respect the behaviour should be the same.

Previously an error would stop any work on the list of files initially gathered regardless of whether the yet to processed files cause failure. This implementation should allow work to proceed on all files gathered initially, but still fail overall in the same way as it currently does on encountering any error.

I don't think this changes the behaviour of the lambda overall however due to the way it's triggered. This lambda is triggered both on a schedule and by events caused by files with specific suffixes being changed in S3. So currently I think that if errors are encountered in an S3 event triggered run the rest of the files that don't get processed will get picked up the scheduled run. 

Either way all the files get processed and failures are reported, but this change means we'll get to some files quicker (that would otherwise have been delayed by an error in their batch).

I also don't think there is anything downstream of this lambda that is concerned with errors emitted. I can't see any alarms connected to failure either which may be something we want to address.

Reply:
that all sounds good then, thanks for the thorough response!